### PR TITLE
Fix RSAPSS salt length conversion

### PR DESCRIPTION
**Pull request by qmuntal**
Fix the conversion between our sentinel salt length variables and the OpenSSL versions in SignRSAPSS and VerifyRSAPSS . We previously set -1 (hash length equals salt length) when 0 was passed when we should've been setting -2 or -3.

This buggy behavior was inherited from BoringSSL bindings since the very beginning, but it hasn't triggered any bug yet because Go `rsa.SignPSS` and `rsa.VerifyPSS` resolved the salt length special case before passing it to boring. There is an [upstream CL](https://go-review.googlesource.com/c/go/+/426659) that will remove this handling from the Go side and fix the Boring side.

We should fix the OpenSSL bindings before that CL lands, else the sync process will fail in the test phase.

### CODE REVIEW

File: openssl/rsa.go [hunk 1]
```diff
@@ -170,18 +170,47 @@ func EncryptRSANoPadding(pub *PublicKeyRSA, msg []byte) ([]byte, error) {
...
+	if golen <= -2 {
+		return 0, errors.New("crypto/rsa: PSSOptions.SaltLength cannot be negative")
+	}
+	clen := C.int(golen)
+	// OpenSSL uses sentinel salt length values like Go do,
+	// but the values don't fully match for rsa.PSSSaltLengthAuto (0).
+	if golen == 0 {
+		if sign {
+			if vMajor == 1 {
+				// OpenSSL 1.x uses -2 to mean maximal size when signing where Go use 0.
```
Comment (dagood) [inline openssl/rsa.go:185]:
```suggestion
				// OpenSSL 1.x uses -2 to mean maximal size when signing where Go crypto uses 0.
```

File: openssl/rsa.go [hunk 2]
```diff
@@ -170,18 +170,47 @@ func EncryptRSANoPadding(pub *PublicKeyRSA, msg []byte) ([]byte, error) {
...
+				clen = C.GO_RSA_PSS_SALTLEN_MAX_SIGN
+			} else {
+				// OpenSSL 3.x deprecated RSA_PSS_SALTLEN_MAX_SIGN
+				// and uses -3 to mean maximal size when signing where Go use 0.
```
Comment (dagood) [inline openssl/rsa.go:189]:
```suggestion
				// and uses -3 to mean maximal size when signing where Go crypto uses 0.
```

File: openssl/rsa.go [hunk 3]
```diff
@@ -170,18 +170,47 @@ func EncryptRSANoPadding(pub *PublicKeyRSA, msg []byte) ([]byte, error) {
...
+				clen = C.GO_RSA_PSS_SALTLEN_MAX
+			}
+		} else {
+			// OpenSSL uses -2 to mean auto-detect size when verifying where Go use 0.
```
Comment (dagood) [inline openssl/rsa.go:193]:
```suggestion
			// OpenSSL uses -2 to mean auto-detect size when verifying where Go crypto uses 0.
```

File: openssl/rsa.go [hunk 4]
```diff
@@ -170,18 +170,47 @@ func EncryptRSANoPadding(pub *PublicKeyRSA, msg []byte) ([]byte, error) {
...
+			clen = C.GO_RSA_PSS_SALTLEN_AUTO
+		}
+	}
+	return clen, nil
+}
+
 func SignRSAPSS(priv *PrivateKeyRSA, h crypto.Hash, hashed []byte, saltLen int) ([]byte, error) {
-	if saltLen == 0 {
-		saltLen = -1 // RSA_PSS_SALTLEN_DIGEST
+	slen, err := saltLength(saltLen, true)
```
Comment (dagood) [inline openssl/rsa.go:201]: This seems like a lot of things that only vary by how many letters of `saltLength` they chop out. 😄 This is what comes to mind:

```suggestion
	opensslSaltLen, err := convertGoSaltLength(saltLen, true)
```

But... these functions are short, and the types are different, so there's not much chance to actually mix them up and I think it's fine as-is if you prefer.

File: openssl/rsa_test.go [hunk 5]
```diff
@@ -118,17 +119,35 @@ func TestSignVerifyPKCS1v15_Invalid(t *testing.T) {
...
 	hashed := sha256.Sum(nil)
-	signed, err := openssl.SignRSAPSS(priv, crypto.SHA256, hashed, 0)
-	if err != nil {
-		t.Fatal(err)
-	}
-	err = openssl.VerifyRSAPSS(pub, crypto.SHA256, hashed, signed, 0)
-	if err != nil {
-		t.Fatal(err)
+	for i, test := range saltLengthCombinations {
+		signed, err := openssl.SignRSAPSS(priv, crypto.SHA256, hashed, test.signSaltLength)
```
Comment (dagood) [inline openssl/rsa_test.go:142]: When I saw this, I thought it was odd that it wasn't set up as a normal table test with named cases, but I see it's based on https://go-review.googlesource.com/c/go/+/426659/4/src/crypto/rsa/pss_test.go. Might be worth linking to it, https://github.com/golang/go/blob/54182ff54a687272dd7632c3a963e036ce03cb7c/src/crypto/rsa/pss_test.go#L200 ?

**Review by dagood — APPROVED** [review]
Functionally LGTM, just a few naming/doc suggestions.

File: openssl/rsa.go (hunk 4 above)
Comment (qmuntal) [reply openssl/rsa.go]: Too mach chopping 🥢. I've simplified `saltLength` and changed a couple of names to make it clearer.

**Review by dagood — APPROVED** [review]

**Review by chsienki — APPROVED** [review]

**Pull request merged**